File: src/main/scala/com/github/mauricio/postgresql/column/TimeEncoderDecoder.scala
package com.github.mauricio.postgresql.column

import org.joda.time.format.DateTimeFormat
import org.joda.time.LocalTime

/**
 * User: Maurício Linhares
 * Date: 3/5/12
 * Time: 6:13 PM
 */

object TimeEncoderDecoder {
  val Instance = new TimeEncoderDecoder()
}

class TimeEncoderDecoder extends ColumnEncoderDecoder {

  private val parser = DateTimeFormat.forPattern("HH:mm:ss.SSSSSS")

  def formatter = parser

  def decode(value: String): LocalTime = {
    parser.parseLocalTime(value)
  }

  override def encode( value : Any ) : String = {
    this.parser.print( value.asInstanceOf[LocalTime] )
  }

}

File: db-async-common/src/main/java/com/github/jasync/sql/db/column/TimeEncoderDecoder.kt
package com.github.jasync.sql.db.column

import com.github.jasync.sql.db.util.microsecondsFormatter
import java.time.LocalTime
import java.time.format.DateTimeFormatterBuilder

open class TimeEncoderDecoder : ColumnEncoderDecoder {
    companion object {
        val Instance = TimeEncoderDecoder()
    }

    private val format = DateTimeFormatterBuilder()
        .appendPattern("HH:mm:ss")
        .appendOptional(microsecondsFormatter)
        .toFormatter()

    private val printer = DateTimeFormatterBuilder()
        .appendPattern("HH:mm:ss.SSSSSS")
        .toFormatter()

    open fun formatter() = format

    override fun decode(value: String): LocalTime =
        LocalTime.parse(value, formatter())

    override fun encode(value: Any): String =
        (value as LocalTime).format(printer)
}
